pkg/storage/iterators.go

```diff
@@ -407,6 +409,149 @@ func NewConditionsFilteredTupleKeyIterator(iter TupleKeyIterator, filter TupleKe
+	lastHead    *openfgav1.Tuple // GUARDED_BY(mu)
+	lastYielded *openfgav1.Tuple // GUARDED_BY(mu)
+}
+
+var _ TupleIterator = (*OrderedCombinedIterator)(nil)
+
+// TODO(miparnisari): this interface & impls belong, conceptually, in internal/graph/tuplemapper.go, but i don't want to have storage import graph.
+type TupleMapper func(t *openfgav1.Tuple) string
+
+func UserMapper() TupleMapper {
+	return func(t *openfgav1.Tuple) string {
+		return t.GetKey().GetUser()
+	}
+}
+
+func ObjectMapper() TupleMapper {
+	return func(t *openfgav1.Tuple) string {
+		return t.GetKey().GetObject()
+	}
+}
+
+// NewOrderedCombinedIterator is a thread-safe iterator that combines a list of iterators into a single ordered iterator.
+// All the input iterators must be individually ordered already according to mapper.
+// Iterators can yield the same value (as defined by mapper) multiple times, but it will only be returned once.
+func NewOrderedCombinedIterator(mapper TupleMapper, sortedIters ...TupleIterator) *OrderedCombinedIterator {
+	pending := make([]TupleIterator, 0, len(sortedIters))
+	for _, sortedIter := range sortedIters {
+		if sortedIter != nil {
+			pending = append(pending, sortedIter)
+		}
+	}
+	return &OrderedCombinedIterator{pending: pending, once: &sync.Once{}, mu: &sync.Mutex{}, mapper: mapper}
+}
+
+func (c *OrderedCombinedIterator) Next(ctx context.Context) (*openfgav1.Tuple, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	idx, err := c.head(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	c.lastHead = nil // invalidate it
+	t, err := c.pending[idx].Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+	c.lastYielded = t
+	return t, nil
+}
+
+func (c *OrderedCombinedIterator) Head(ctx context.Context) (*openfgav1.Tuple, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.lastHead != nil {
+		return c.lastHead, nil
+	}
+	idx, err := c.head(ctx)
+	if err != nil {
+		return nil, err
+	}
+	lastHead, err := c.pending[idx].Head(ctx)
+	c.lastHead = lastHead
+	return c.lastHead, err
+}
+
+// head returns the index (within the pending array) that has the next smallest element.
+// The pending array is mutated, and there may be nil elements in it after this function runs.
+// Callers must use the returned index immediately, without mutating the pending array.
+// NOTE: callers must hold mu.
+func (c *OrderedCombinedIterator) head(ctx context.Context) (int, error) {
+	c.clearPendingThatAreNil()
+
+	var headMin *openfgav1.Tuple
+	minIdx := -1
+
+IterateOverPending:
+	for pendingIdx, iter := range c.pending {
+		head, err := iter.Head(ctx)
+		if err != nil {
+			if errors.Is(err, ErrIteratorDone) {
+				iter.Stop()
+				c.pending[pendingIdx] = nil
+				continue
+			}
+			return minIdx, err
+		}
+
+		if c.lastYielded != nil {
+			if c.mapper(head) < c.mapper(c.lastYielded) {
+				return minIdx, fmt.Errorf("iterator %d is not in ascending order", pendingIdx)
+			}
+			// discard duplicate values
+			for c.mapper(head) == c.mapper(c.lastYielded) {
```

Comment:
Curious why we choose to prune at the next head() vs. in the current head()?  In this way, we don't have to remember lastYielded and makes the code easier to understand.

Reply:
It's been a few weeks since I wrote this code :( IIRC i did it the way you suggested originally but @elbuo8 suggested the current implementation because it's faster.

Reply:
Great.  Maybe add comments in the code to point out this is done for performance as it is not obvious at first.